packages/celotool/src/e2e-tests/utils.ts

```diff
@@ -385,109 +125,155 @@ export async function migrateContracts(
+      await checkoutGethRepo(branch, gethConfig.gethRepoPath)
+    }
+
+    if (gethConfig.gethRepoPath) {
+      await buildGeth(gethConfig.gethRepoPath)
+    }
+
+    if (!gethConfig.keepData && fs.existsSync(gethConfig.runPath)) {
+      await resetDataDir(gethConfig.runPath, verbose)
+    }
+
+    if (!fs.existsSync(gethConfig.runPath)) {
+      // @ts-ignore
+      fs.mkdirSync(gethConfig.runPath, { recursive: true })
     }
-    await buildGeth(gethRepoPath)
-    await setupTestDir(TEST_DIR)
-    await writeGenesis(validators, GENESIS_PATH, gethConfig.genesisConfig)
+
+    await writeGenesis(gethConfig, validators, verbose)
+
+    let bootnodeEnode: string = ''
+
+    if (gethConfig.useBootnode) {
+      bootnodeEnode = await startBootnode(bootnodeBinaryPath, mnemonic, gethConfig, verbose)
+    }
+
     let validatorIndex = 0
+    let proxyIndex = 0
+
     for (const instance of gethConfig.instances) {
+      // Non proxied validators and proxies should connect to the bootnode
+      if ((instance.validating && !instance.isProxied) || instance.isProxy) {
+        if (gethConfig.useBootnode) {
+          instance.bootnodeEnode = bootnodeEnode
+        }
+      } else if (instance.validating && instance.isProxied) {
+        // Proxied validators should connect to only the proxy
+        // Find this proxied validator's proxy
+        const proxyEnode = proxyEnodes.filter((x: any) => x[0] === instance.proxy)
+
+        if (proxyEnode.length !== 1) {
+          throw new Error('proxied validator must have exactly one proxy')
+        }
+
+        instance.proxies = [proxyEnode[0][1]!, proxyEnode[0][2]!]
+      }
+
+      // Set the private key for the validator or proxy instance
       if (instance.validating) {
         instance.privateKey = instance.privateKey || validatorPrivateKeys[validatorIndex]
         validatorIndex++
+      } else if (instance.isProxy) {
+        instance.privateKey = instance.privateKey || proxyPrivateKeys[proxyIndex]
+        proxyIndex++
+      }
+    }
+
+    // The proxies will need to know their proxied validator's address
+    for (const instance of gethConfig.instances) {
+      if (instance.isProxy) {
+        const proxiedValidator = gethConfig.instances.filter(
+          (x: GethInstanceConfig) => x.proxy === instance.name
+        )
+
+        if (proxiedValidator.length !== 1) {
+          throw new Error('proxied validator must have exactly one proxy')
+        }
+
+        instance.proxiedValidatorAddress = privateKeyToAddress(proxiedValidator[0].privateKey!)
       }
-      await initAndStartGeth(gethBinaryPath, instance)
     }
-    await connectValidatorPeers(gethConfig)
+
+    // Start all the instances
+    for (const instance of gethConfig.instances) {
+      await initAndStartGeth(gethConfig, gethBinaryPath, instance, verbose)
+    }
+
+    await connectValidatorPeers(gethConfig.instances)
+
+    // Give validators time to connect to each other
+    await sleep(60, true)
```

Comment:
A full minute is probably too much and will likely become annoying with time. I'd either recommend just lowering this value (migration can afford to start a little early because the transactions will propagate when the nodes connect) or, for extra credit, finding some way to poll the nodes to see when they are connected and wait for that to be true.

Reply:
replaced with `waitForSync` methods